File: BrewPOS-Workspace/brewpos-backend/src/routes/menu.ts
import { Router } from 'express';
import { prisma } from '../db';
import { logActivity } from '../services/systemLogger';
import multer from 'multer';
import path from 'path';
import fs from 'fs';

const router = Router();

// Ensure uploads directory exists
const uploadDir = path.join(__dirname, '../../public/uploads');
if (!fs.existsSync(uploadDir)) {
  fs.mkdirSync(uploadDir, { recursive: true });
}

const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, uploadDir);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
    cb(null, uniqueSuffix + path.extname(file.originalname));
  }
});
const upload = multer({ storage });

// Get all menus
router.get('/', async (req, res) => {
  try {
    const itemQuantities = await prisma.transactionItem.groupBy({
      by: ['menuId'],
      _sum: {
        quantity: true
      }
    });
    
    const quantityMap = new Map(itemQuantities.map(item => [item.menuId, item._sum.quantity || 0]));

    const menus = await prisma.menu.findMany({
      include: { category: true },
    });
    
    const menusWithSoldCount = menus.map(menu => ({
      ...menu,
      soldCount: quantityMap.get(menu.id) || 0
    }));

    res.json(menusWithSoldCount);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch menus' });
  }
});

// Create menu
router.post('/', upload.single('image'), async (req, res) => {
  try {
    const { name, description, price, categoryId, hpp } = req.body;
    let imageUrl = null;
    
    if (req.file) {
      imageUrl = `/uploads/${req.file.filename}`;
    } else if (req.body.imageUrl) {
      imageUrl = req.body.imageUrl;
    }

    const menu = await prisma.menu.create({
      data: {
        name,
        description,
        price: Number(price),
        hpp: hpp !== undefined ? Number(hpp) : 0,
        categoryId: Number(categoryId),
        imageUrl
      },
    });
    
    await logActivity('CREATE_MENU', req.headers['x-user-id'] ? Number(req.headers['x-user-id']) : undefined, `Created menu: ${name}`, 'Menu', menu.id);
    
    res.status(201).json(menu);
  } catch (error) {
    console.error('Error creating menu:', error);
    res.status(500).json({ error: 'Failed to create menu', details: error instanceof Error ? error.message : String(error) });
  }
});

// Update menu
router.put('/:id', upload.single('image'), async (req, res) => {
  try {
    const { id } = req.params;
    const { name, description, price, categoryId, hpp } = req.body;
    
    // Find existing menu to check current image if no new one provided
    const existingMenu = await prisma.menu.findUnique({ where: { id: Number(id) } });
    if (!existingMenu) {
      return res.status(404).json({ error: 'Menu not found' });
    }

    let imageUrl = existingMenu.imageUrl;
    if (req.file) {
      imageUrl = `/uploads/${req.file.filename}`;
    } else if (req.body.imageUrl !== undefined) {
      imageUrl = req.body.imageUrl;
    }

    const menu = await prisma.menu.update({
      where: { id: Number(id) },
      data: {
        name,
        description,
        price: Number(price),
        hpp: hpp !== undefined ? Number(hpp) : existingMenu.hpp,
        categoryId: Number(categoryId),
        imageUrl
      },
    });

    await logActivity('UPDATE_MENU', req.headers['x-user-id'] ? Number(req.headers['x-user-id']) : undefined, `Updated menu: ${menu.name}`, 'Menu', menu.id);

    res.json(menu);
  } catch (error) {
    console.error('Error updating menu:', error);
    res.status(500).json({ error: 'Failed to update menu', details: error instanceof Error ? error.message : String(error) });
  }
});

// Delete menu
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    
    // Check if menu exists
    const menu = await prisma.menu.findUnique({ where: { id: Number(id) } });
    if (!menu) {
      return res.status(404).json({ error: 'Menu not found' });
    }

    // Try deleting the menu
    // If it's used in transactions, Prisma will throw a foreign key constraint error
    await prisma.menu.delete({
      where: { id: Number(id) },
    });

    await logActivity('DELETE_MENU', req.headers['x-user-id'] ? Number(req.headers['x-user-id']) : undefined, `Deleted menu: ${menu.name}`, 'Menu', menu.id);

    res.status(204).send();
  } catch (error: any) {
    console.error('Error deleting menu:', error);
    // Prisma Foreign Key Constraint Failed Error Code: P2003
    if (error.code === 'P2003') {
      res.status(400).json({ error: 'Cannot delete menu because it is already used in transactions. Please hide or disable it instead (feature coming soon).' });
    } else {
      res.status(500).json({ error: 'Failed to delete menu', details: error.message });
    }
  }
});

export default router;

File: BrewPOS-Workspace/brewpos-flask/app/routes/menu.py
import os
import time
from flask import Blueprint, request, jsonify, current_app
from app.middleware.auth import get_current_user_id
from werkzeug.utils import secure_filename
from sqlalchemy import func
from app.extensions import db
from app.models.menu import Menu, RecipeIngredient
from app.models.transaction import TransactionItem
from app.models.inventory import InventoryItem
from app.services.system_logger import log_activity

menu_bp = Blueprint('menu', __name__, url_prefix='/api/menus')

def _save_upload_file(file):
    if not file or file.filename == '':
        return None
    filename = secure_filename(f"menu_{int(time.time()*1000)}_{file.filename}")
    upload_dir = os.path.join(current_app.config['UPLOAD_FOLDER'])
    os.makedirs(upload_dir, exist_ok=True)
    file_path = os.path.join(upload_dir, filename)
    file.save(file_path)
    return f"/uploads/{filename}"

@menu_bp.route('', methods=['GET'])
def get_menus():
    try:
        # Sum quantity sold per menu
        sold_counts = db.session.query(
            TransactionItem.menuId,
            func.sum(TransactionItem.quantity).label('total_sold')
        ).group_by(TransactionItem.menuId).all()

        sold_map = {sc[0]: int(sc[1] or 0) for sc in sold_counts}

        # POS memanggil dengan activeOnly=1; halaman kelola menu tetap melihat
        # yang diarsipkan supaya bisa diaktifkan kembali.
        query = Menu.query
        if request.args.get('activeOnly') in ('1', 'true'):
            # MSSQL menolak "IS 1" yang dihasilkan is_(True); pakai perbandingan biasa.
            query = query.filter(Menu.isActive == True)  # noqa: E712

        menus = query.all()
        result = [m.to_dict(include_category=True, sold_count=sold_map.get(m.id, 0)) for m in menus]
        return jsonify(result)
    except Exception as e:
        print(f"Error fetching menus: {e}")
        return jsonify({'error': 'Failed to fetch menus', 'details': str(e)}), 500


@menu_bp.route('', methods=['POST'])
def create_menu():
    try:
        # Support both multipart form data and json
        if request.is_json:
            data = request.get_json() or {}
            name = data.get('name')
            description = data.get('description')
            price = int(data.get('price', 0))
            hpp = int(data.get('hpp', 0))
            category_id = int(data.get('categoryId', 0))
            image_url = data.get('imageUrl')
        else:
            name = request.form.get('name')
            description = request.form.get('description')
            price = int(request.form.get('price', 0))
            hpp = int(request.form.get('hpp', 0))
            category_id = int(request.form.get('categoryId', 0))
            image_url = request.form.get('imageUrl')
            
            if 'image' in request.files:
                uploaded_url = _save_upload_file(request.files['image'])
                if uploaded_url:
                    image_url = uploaded_url

        if not name or not price or not category_id:
            return jsonify({'error': 'Name, price, and categoryId are required'}), 400

        menu = Menu(
            name=name,
            description=description,
            price=price,
            hpp=hpp,
            categoryId=category_id,
            imageUrl=image_url
        )
        db.session.add(menu)
        db.session.commit()

        user_id = get_current_user_id()
        log_activity('CREATE_MENU', int(user_id) if user_id else None,
                     f"Created menu: {name}", 'Menu', menu.id)

        return jsonify(menu.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        print(f"Error creating menu: {e}")
        return jsonify({'error': 'Failed to create menu', 'details': str(e)}), 500


@menu_bp.route('/<int:id>', methods=['PUT'])
def update_menu(id):
    try:
        menu = Menu.query.get(id)
        if not menu:
            return jsonify({'error': 'Menu not found'}), 404

        if request.is_json:
            data = request.get_json() or {}
            if 'name' in data: menu.name = data['name']
            if 'description' in data: menu.description = data['description']
            if 'price' in data: menu.price = int(data['price'])
            if 'hpp' in data: menu.hpp = int(data['hpp'])
            if 'categoryId' in data: menu.categoryId = int(data['categoryId'])
            if 'imageUrl' in data: menu.imageUrl = data['imageUrl']
            if 'isActive' in data: menu.isActive = bool(data['isActive'])
        else:
            if 'name' in request.form: menu.name = request.form['name']
            if 'description' in request.form: menu.description = request.form['description']
            if 'price' in request.form: menu.price = int(request.form['price'])
            if 'hpp' in request.form: menu.hpp = int(request.form['hpp'])
            if 'categoryId' in request.form: menu.categoryId = int(request.form['categoryId'])
            if 'imageUrl' in request.form: menu.imageUrl = request.form['imageUrl']

            if 'image' in request.files:
                uploaded_url = _save_upload_file(request.files['image'])
                if uploaded_url:
                    menu.imageUrl = uploaded_url

        db.session.commit()

        user_id = get_current_user_id()
        log_activity('UPDATE_MENU', int(user_id) if user_id else None,
                     f"Updated menu: {menu.name}", 'Menu', menu.id)

        return jsonify(menu.to_dict())
    except Exception as e:
        db.session.rollback()
        print(f"Error updating menu: {e}")
        return jsonify({'error': 'Failed to update menu', 'details': str(e)}), 500


@menu_bp.route('/<int:id>', methods=['DELETE'])
def delete_menu(id):
    try:
        menu = Menu.query.get(id)
        if not menu:
            return jsonify({'error': 'Menu not found'}), 404

        # Check if used in transactions
        used_count = TransactionItem.query.filter_by(menuId=id).count()
        if used_count > 0:
            return jsonify({
                'error': f'Menu "{menu.name}" sudah terpakai di {used_count} transaksi, '
                         'jadi tidak bisa dihapus tanpa merusak riwayat penjualan. '
                         'Nonaktifkan saja agar hilang dari POS.',
                'canArchive': True,
                'menuId': menu.id,
                'usedCount': used_count
            }), 400

        name = menu.name
        db.session.delete(menu)
        db.session.commit()

        user_id = get_current_user_id()
        log_activity('DELETE_MENU', int(user_id) if user_id else None,
                     f"Deleted menu: {name}", 'Menu', id)

        return '', 204
    except Exception as e:
        db.session.rollback()
        print(f"Error deleting menu: {e}")
        return jsonify({'error': 'Failed to delete menu', 'details': str(e)}), 500


# --- RECIPE INGREDIENTS MANAGEMENT ---

@menu_bp.route('/<int:id>/recipe', methods=['GET'])
def get_menu_recipe(id):
    try:
        menu = Menu.query.get(id)
        if not menu:
            return jsonify({'error': 'Menu not found'}), 404

        ingredients = RecipeIngredient.query.filter_by(menuId=id).all()
        return jsonify([i.to_dict() for i in ingredients])
    except Exception as e:
        return jsonify({'error': 'Failed to fetch menu recipe', 'details': str(e)}), 500


@menu_bp.route('/<int:id>/recipe', methods=['POST'])
def set_menu_recipe(id):
    """
    Sets the full recipe for a menu.
    Payload: { "ingredients": [ { "inventoryItemId": 1, "quantityNeeded": 18.0 }, ... ] }
    """
    try:
        menu = Menu.query.get(id)
        if not menu:
            return jsonify({'error': 'Menu not found'}), 404

        data = request.get_json() or {}
        ingredients_data = data.get('ingredients', [])

        # Remove existing recipe ingredients and replace with new ones
        RecipeIngredient.query.filter_by(menuId=id).delete()

        created_items = []
        for item in ingredients_data:
            inv_id = item.get('inventoryItemId')
            qty = float(item.get('quantityNeeded', 0))
            if inv_id and qty > 0:
                inv_item = InventoryItem.query.get(inv_id)
                if inv_item:
                    recipe_ing = RecipeIngredient(
                        menuId=id,
                        inventoryItemId=inv_id,
                        quantityNeeded=qty
                    )
                    db.session.add(recipe_ing)
                    created_items.append(recipe_ing)

        db.session.commit()
        return jsonify([i.to_dict() for i in created_items]), 200
    except Exception as e:
        db.session.rollback()
        print(f"Error setting menu recipe: {e}")
        return jsonify({'error': 'Failed to save menu recipe', 'details': str(e)}), 500


@menu_bp.route('/<int:id>/recipe/<int:recipe_id>', methods=['DELETE'])
def delete_recipe_ingredient(id, recipe_id):
    try:
        recipe_ing = RecipeIngredient.query.filter_by(id=recipe_id, menuId=id).first()
        if not recipe_ing:
            return jsonify({'error': 'Recipe ingredient not found'}), 404

        db.session.delete(recipe_ing)
        db.session.commit()
        return jsonify({'message': 'Ingredient removed from recipe'})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to delete recipe ingredient', 'details': str(e)}), 500
